os/binfmt/libelf/libelf_cache.c

```diff
@@ -0,0 +1,494 @@
+		readsize = file_len % cache_blocks_size;
+	} else {
+		readsize = cache_blocks_size;
+	}
+
+	/* Read actual data to 'block_number's buf */
+	nbytes = read(filfd, buf, readsize);
+	if (nbytes != readsize) {
+		berr("Read for actual elf block# %d of size %d failed\n", block_number, readsize);
+		return ERROR;
+	}
+
+	return nbytes;
+}
+
+/****************************************************************************
+ * Name: libelf_cache_update_blockcache_list
+ *
+ * Description:
+ *   Update blockcache list based on whether 'block_number' block in
+ *   blockwise-elf binary is already cached or not. Also, maintain
+ *   head, tail and most_accessed pointers.
+ *   tail = most recently cached block (highest priority block for keeping
+ *          cached)
+ *   head = oldest cached block (lowest priority for keeping it cached).
+ *          If a new block needs to be cached, cache to this blockcache element.
+ *   most_accessed = pointer to most accessed block in blockcache list.
+ *          Needed for easy access to most accessed block from comp. binary.
+ *
+ * Returned Value:
+ *   Index in blockcache list where 'block_number' from elf
+ *   binary is cached state. Return Negative value on failure.
+ ****************************************************************************/
+static unsigned int libelf_cache_update_blockcache_list(int block_number, int filfd, uint16_t binary_header_size)
+{
+	unsigned int blockcache_index;		/* Which blockcache element has needed ELF data for read */
+	block_cache_t *ptr;					/* Pointer to element in blockcache list which will be updated */
+	block_cache_t *temp;				/* Used when updating blockcache list */
+	bool flag_for_caching;				/* 1 = Block needs to be cached, 0 = Block already cached */
+	bool update_most_accessed;			/* 1 = Most accessed block is being removed from blockcache.
+										 *     Need to update most accessed pointer from scratch */
+	int size;
+
+	binfo("filfd: %d block_number: %d\n", filfd, block_number);
+
+	/* Initialize flags */
+	flag_for_caching = true;
+	update_most_accessed = false;
+
+	/* First check is with block_number at most_accessed pointer */
+	if (most_accessed->block_number == block_number) {
+		flag_for_caching = false;
+		blockcache_index = most_accessed->index_block_cache;
+	}
+
+	/* Start checking for if block is cached from the tail */
+	if (flag_for_caching == true) {
+		ptr = tail;
+		while (ptr != NULL) {
+			if (ptr->block_number == block_number) {
+				flag_for_caching = false;
+				blockcache_index = ptr->index_block_cache;
+				break;
+			} else {
+				ptr = ptr->prev;
+			}
+		}
+	}
+
+	/* If block_number block from elf binary needs to be cached */
+	if (flag_for_caching == true) {
+
+		/* If most_accessed == head, need to update most_accessed */
+		if (head == most_accessed)
+			update_most_accessed = true;
+
+		/* Detach head, Reassign head to head->next */
+		ptr = head;
+		head = ptr->next;
+		head->prev = ptr->prev;
+
```

Comment:
While detaching head, i think head pointer next & prev should be reassigned.

Reply:
Taken care in elf_cache_unint function